Mathlib/Algebra/Ring/SumsOfSquares.lean

```diff
@@ -0,0 +1,128 @@
+
+- The type `SumSqIn R` : in an additive monoid with multiplication `R`, we introduce the type
+`SumSqIn R` as the submonoid of `R` whose carrier is the subset `{S : R | isSumSq S}`.
+
+## Theorems
+
+- `isSumSq.add`: if `S1` and `S2` are sums of squares in `R`, then so is `S1 + S2`.
+- `SumSq.nonneg`: in a linearly ordered semiring `R` with an `[ExistsAddOfLE R]` instance, sums of
+squares are non-negative.
+- `SquaresInSumSq`: a square is a sum of squares.
+- `SquaresAddClosure`: the submonoid of `R` generated by the squares in `R` is the set of sums of
+squares in `R`.
+
+-/
+
+variable {R : Type*}
+
+/--
+In a type `R` with an addition, a zero element and a multiplication, the property of being a sum of
+squares is defined by an inductive predicate: `0 : R` is a sum of squares and if `S` is a sum of
+squares, then for all `a : R`, `a * a + S` is a sum of squares in `R`.
+-/
+
+@[mk_iff]
+inductive isSumSq [Add R] [Zero R] [Mul R] : R → Prop
+  | zero                              : isSumSq 0
+  | sq_add (a S : R) (pS : isSumSq S) : isSumSq (a * a + S)
+
+/--
+If `S1` and `S2` are sums of squares in a semiring `R`, then `S1 + S2` is a sum of squares in `R`.
+-/
+
+theorem isSumSq.add [AddMonoid R] [Mul R] {S1 S2 : R} (p1 : isSumSq S1)
+    (p2 : isSumSq S2) : isSumSq (S1 + S2) := by
+  induction p1 with
+  | zero             => rewrite [zero_add]; exact p2
+  | sq_add a S pS ih => rewrite [add_assoc]; exact isSumSq.sq_add a (S + S2) ih
+
+variable (R) in
+
+/--
+In an additive monoid with multiplication `R`, the type `SumSqIn R` is the submonoid of sums of
+squares in `R`.
+-/
+
+def SumSqIn [AddMonoid R] [Mul R] : AddSubmonoid R where
+  carrier   := {S : R | isSumSq S}
+  zero_mem' := isSumSq.zero
+  add_mem'  := isSumSq.add
+
+variable (R) in
+
+/--
+The set of squares in a type `R` endowed with a `Mul R` instance. By definition, a square in `R` is
+a term `x : R` such that `x = y * y` for some `y : R` (see Mathlib.Algebra.Group.Even).
+-/
+
+def Squares [Mul R] : Set R := {x : R | IsSquare x}
+
+/--
+In an additive monoid with multiplication, every square is a sum of squares.
+-/
+
+theorem SquaresInSumSq [AddMonoid R] [Mul R ] : Squares R ⊆ SumSqIn R := by
+  intro x px
+  rcases px with ⟨y, py⟩
+  simp [SumSqIn]
+  rw [py, ← AddMonoid.add_zero (y * y)]
+  apply isSumSq.sq_add
+  exact isSumSq.zero
+
+/--
+In an additive monoid with multiplication `R`, the submonoid generated by the squares is the set of
+sums of squares.
+-/
+
+theorem SquaresAddClosure [AddMonoid R] [Mul R] : AddSubmonoid.closure (Squares R) = SumSqIn R := by
+  ext x
+  constructor
+  · intro px
+    refine AddSubmonoid.mem_carrier.mp ?h.mp.a
+    simp only [AddSubmonoid.mem_closure] at px
+    apply px (SumSqIn R)
+    exact SquaresInSumSq
+  · intro px
+    apply AddSubmonoid.mem_closure.mpr
+    intro M pM
+    induction' px with a S _ pS
+    · exact AddSubmonoid.zero_mem M
+    · refine AddSubmonoid.add_mem M ?h.mpr.sq_add.a pS
+      apply pM
+      use a
```

Comment:
```suggestion
  refine le_antisymm (closure_le.2 (fun x hx ↦ SquaresInSumSq hx)) (fun x hx ↦ ?_)
  induction hx with
  | zero             => exact zero_mem _
  | sq_add a S _ ih  => exact AddSubmonoid.add_mem _ (subset_closure ⟨a, rfl⟩) ih
```
This proof is shorter, but I also think it closer to what we do mathematically: `SumSqIn` is a submnoid, so by the universal property of the closure to prove that `closure IsSquare` is included in `SumSqIn` is it enough to prove that `IsSquare` is in `SumSqIn`, and you have already proved it. The proof of the other inclusion is basically the same as yours, but we try to avoid using `induction'`, that sooner or later will disappear.

Reply:
This one does not quite work for me, I think I may be missing something regarding where you want me to use this (I assume in `SquaresAddClosure`).

Reply:
No, but I added the prefix. I'll try again.

Reply:
OK, got it!